File: server/internal/handler/recently_played.go
package handler

import (
	"context"
	"encoding/json"
	"net/http"
	"strconv"

	"github.com/mcopland/spotifind/internal/middleware"
	"github.com/mcopland/spotifind/internal/models"
)

// RecentlyPlayedLister is satisfied by repository.RecentlyPlayedRepo.
type RecentlyPlayedLister interface {
	ListForUser(ctx context.Context, userID int64, f models.RecentlyPlayedFilters) (*models.PaginatedResult[models.RecentlyPlayedTrack], error)
}

type RecentlyPlayedHandler struct {
	repo RecentlyPlayedLister
}

func NewRecentlyPlayedHandler(repo RecentlyPlayedLister) *RecentlyPlayedHandler {
	return &RecentlyPlayedHandler{repo: repo}
}

func (h *RecentlyPlayedHandler) List(w http.ResponseWriter, r *http.Request) {
	userID, ok := middleware.GetUserID(r.Context())
	if !ok {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return
	}

	q := r.URL.Query()
	f := models.RecentlyPlayedFilters{}
	if v := q.Get("page"); v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			f.Page = n
		}
	}
	if v := q.Get("page_size"); v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			f.PageSize = n
		}
	}

	result, err := h.repo.ListForUser(r.Context(), userID, f)
	if err != nil {
		http.Error(w, "failed to list recently played", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(result)
}

File: web/src/api/recently-played.ts
import type { PaginatedResult, RecentlyPlayedTrack } from "../types";
import client from "./client";

export async function getRecentlyPlayed(params: {
  page?: number;
  page_size?: number;
}): Promise<PaginatedResult<RecentlyPlayedTrack>> {
  const p: Record<string, string> = {};
  if (params.page) p.page = String(params.page);
  if (params.page_size) p.page_size = String(params.page_size);
  const res = await client.get<PaginatedResult<RecentlyPlayedTrack>>("/recently-played", { params: p });
  return res.data;
}
